src/typechecker/FStar.TypeChecker.TcInductive.fs

```diff
@@ -1034,8 +1034,8 @@ let check_inductive_well_typedness (env:env_t) (ses:list<sigelt>) (quals:list<qu
                   (univs, body)
               in
               if List.length univs = List.length (fst expected_typ)
-              then let _, inferred = Env.inst_tscheme inferred_typ in
-                   let _, expected = Env.inst_tscheme expected_typ in
+              then let _, inferred = Subst.open_univ_vars univs (snd inferred_typ) in
+                   let _, expected = Subst.open_univ_vars univs (snd expected_typ) in
```

Comment:
`inst_tscheme` is wrong in this context.

Consider what happens if the expected type is `Type u#a -> Type0` while the inferred type is `Type u#a -> Type u#a`. In this case, instantiating the types produces `Type u#?u1 -> Type0` and `Type u#?u2 -> Type u#?u2`. And these two types are unifiable by setting `?u1 = ?u2 = 0`.

Instead, we need to open the type schemes with names (not uvars) and then compare them.